traders/fsm.py

```diff
@@ -0,0 +1,170 @@
+                '\'next_state\' must be an instance of {}, not {}.'
+                .format(State, next_state.__class__))
+
+        if update is not None and not callable(update):
+            raise TypeError('\'update\' must be callable or None, not {}.'
+                            .format(update.__class__))
+
+        self._current_state = current_state
+        self._event = event
+        self._next_state = next_state
+        self._update = update
+
+    def __str__(self):
+        return '({}, {}) -> {}'.format(str(self.current_state),
+                                       str(self.event),
+                                       str(self.next_state))
+
+    @property
+    def event(self):
+        return self._event
+
+    @property
+    def current_state(self):
+        return self._current_state
+
+    @property
+    def next_state(self):
+        """Returns the next state without updating it"""
+        return self._next_state
+
+    def apply(self):
+        """Applies the update function and returns the new state"""
+        if self._update is not None:
+            self._update(self.current_state, self.event, self._next_state)
+        return self._next_state
+
+    def match(self, current_state, event):
+        """Verifies if the transition matches the state and event"""
+
+        if self.current_state == current_state and self.event == event:
+            return True
+
+        return False
+
+
+class FiniteStateMachine(State):
+
+    def __init__(self, name, start_state):
+        """A finite state machine
+
+        The finite state machine encapsulates a list of states and the
+        transition between them. Because a FSM is itself a state, it can be
+        run and yield events. The events of a FSM are all the events of its
+        state which are not associated with a transition.
+
+        """
+
+        super().__init__(name)
+        self.start_state = start_state
+        self._transitions = []
+
+    @property
+    def events(self):
+        """Returns the list of events that can be thrown by the  FSM"""
+
+        # The events that the FSM can throw are all the events that
+        # do not have a transition.
+        events = []
+        states = set(chain((self.start_state,),
+                           (t.current_state for t in self._transitions),
+                           (t.next_state for t in self._transitions)))
+        for state in states:
+            for event in state.events:
+                if self.transition(state, event) is None:
+                    events.append(event)
+
+        return tuple(events)
+
+    def add(self, transition):
+        """Adds a transition to a FiniteStateMachine"""
+
+        if not isinstance(transition, Transition):
+            raise TypeError(
+                '\'transition\' must be an instance of {}, not {}.'
+                .format(Transition, transition.__class__))
+        self._transitions.append(transition)
+
+    def pretty_print(self):
```

Comment:
if it's only for debugging then we should call it _pretty_print, otherwise this can be the string representation